File: javascript/solutions/TennisGame1.js
var TennisGame1 = function (player1Name, player2Name) {
    this.m_score1 = 0;
    this.m_score2 = 0;
    this.player1Name = player1Name;
    this.player2Name = player2Name;
};

TennisGame1.prototype.wonPoint = function (playerName) {
    if (playerName === this.player1Name)
        this.m_score1 += 1;
    else
        this.m_score2 += 1;
};

var ScoreToWords = function (score) {
    return [
        "Love",
        "Fifteen",
        "Thirty",
        "Forty"
    ][score]
}

var isEvenScore = function (score1, score2) {
    return score1 === score2
}

var isAdvantageScore = function (score1, score2) {
    return score1 >= 4 || score2 >= 4
}

var GetEvenScoreInWords = function (score) {
    if (score >= 3) {
        return "Deuce"
    }
    return ScoreToWords(score) + "-All"
}

var GetAdvantageScoreInWords = function (minusResult) {
    if (minusResult === 1) return "Advantage player1";
    else if (minusResult === -1) return "Advantage player2";
    else if (minusResult >= 2) return "Win for player1";
    else return "Win for player2";
}

var GetScoreInWords = function (score1, score2) {
    return ScoreToWords(score1) + "-" + ScoreToWords(score2)
}

TennisGame1.prototype.getScore = function () {
    if (isEvenScore(this.m_score1, this.m_score2)) {
        return GetEvenScoreInWords(this.m_score1)
    }
    if (isAdvantageScore(this.m_score1, this.m_score2)) {
        return GetAdvantageScoreInWords(this.m_score1 - this.m_score2)
    }
    return GetScoreInWords(this.m_score1, this.m_score2)
};

if (typeof window === "undefined") {
    module.exports = TennisGame1;
}

File: csharp/Tennis/TennisGame1.cs
using System.Collections.Generic;

namespace Tennis
{
    class TennisGame1 : ITennisGame
    {
        private Player player1;
        private Player player2;
        private bool isEvenScore { get { return player1.score == player2.score; } }
        private bool isEndGameScore { get { return player1.score >= 4 || player2.score >= 4; } }

        public TennisGame1(string player1Name, string player2Name)
        {
            this.player1 = new Player(player1Name);
            this.player2 = new Player(player2Name);
        }

        private Player GetPlayer(string playerName)
        {
            return playerName == player1.name ? player1 : player2;
        }

        public void WonPoint(string playerName)
        {
            GetPlayer(playerName).WonPoint();
        }

        private string GetEvenScore()
        {
            if (player1.score <= 2)
            {
                return $"{Score[player1.score]}-All";
            }

            return "Deuce";
        }

        private string GetEndGameScore()
        {
            var minusResult = player1.score - player2.score;
            if (minusResult == 1) return $"Advantage {player1.name}";
            else if (minusResult == -1) return $"Advantage {player2.name}";
            else if (minusResult >= 2) return $"Win for {player1.name}";
            return $"Win for {player2.name}";
        }

        private Dictionary<int, string> Score = new Dictionary<int, string>(){
            {0, "Love"},
            {1, "Fifteen"},
            {2, "Thirty"},
            {3, "Forty"},
        };

        private string GetNormalScore()
        {
            return $"{Score[player1.score]}-{Score[player2.score]}";
        }

        public string GetScore()
        {
            if (isEvenScore)
            {
                return GetEvenScore();
            }
            if (isEndGameScore)
            {
                return GetEndGameScore();
            }
            return GetNormalScore();
        }
    }
}

